tests/e2e/main_test.go

```diff
@@ -291,4 +295,104 @@ func TestGrafana(t *testing.T){
+
+// TestDeprecationLogs tests if all components of kube-prometheus doesn't emit deprecation warning in its logs.
+// Kube-prometheus aims to keep itself up-to-date with upstream configuration, which means we don't use deprecated flags.
+func TestDeprecationLogs(t *testing.T) {
+	kClient := promClient.kubeClient
+
+	type getLabelSelectorFunc func(string) labels.Set
+
+	daemonSetGetLabelSelector := func(app string) labels.Set {
+		daemonset, err := kClient.AppsV1().DaemonSets("monitoring").Get(context.Background(), app, metav1.GetOptions{})
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		return labels.Set(daemonset.Spec.Selector.MatchLabels)
+	}
+
+	deploymentGetLabelSelector := func(app string) labels.Set {
+		deployment, err := kClient.AppsV1().Deployments("monitoring").Get(context.Background(), app, metav1.GetOptions{})
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		return labels.Set(deployment.Spec.Selector.MatchLabels)
+	}
+
+	statefulSetGetLabelSelector := func(app string) labels.Set {
+		statefulSet, err := kClient.AppsV1().StatefulSets("monitoring").Get(context.Background(), app, metav1.GetOptions{})
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		return labels.Set(statefulSet.Spec.Selector.MatchLabels)
+	}
+
+	testCases := []struct {
+		app              string
+		container        string
+		getLabelSelector getLabelSelectorFunc
+	}{
+		{
+			app:              "node-exporter",
+			container:        "node-exporter",
+			getLabelSelector: daemonSetGetLabelSelector,
+		}, {
+			app:              "prometheus-k8s",
+			container:        "prometheus",
+			getLabelSelector: statefulSetGetLabelSelector,
+		}, {
+			app:              "alertmanager-main",
+			container:        "alertmanager",
+			getLabelSelector: statefulSetGetLabelSelector,
+		}, {
+			app:              "grafana",
+			container:        "grafana",
+			getLabelSelector: deploymentGetLabelSelector,
+		}, {
+			app:              "kube-state-metrics",
+			container:        "kube-state-metrics",
+			getLabelSelector: deploymentGetLabelSelector,
+		}, {
+			app:              "prometheus-operator",
+			container:        "prometheus-operator",
+			getLabelSelector: deploymentGetLabelSelector,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.app, func(t *testing.T) {
+			pods, err := kClient.CoreV1().Pods("monitoring").List(context.Background(), metav1.ListOptions{LabelSelector: tc.getLabelSelector(tc.app).AsSelector().String()})
+			if err != nil {
+				t.Fatal(err)
+			}
+			podLogOpts := v1.PodLogOptions{
+				Container: tc.container,
+			}
+
+			for _, pod := range pods.Items {
+				req := kClient.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, &podLogOpts)
+				podLogs, err := req.Stream(context.Background())
+				if err != nil {
+					t.Fatal(err)
+				}
+				defer podLogs.Close()
+
+				buf := new(bytes.Buffer)
+				_, err = io.Copy(buf, podLogs)
+				if err != nil {
+					t.Fatal(err)
+				}
+				logs := buf.String()
+				if strings.Contains(strings.ToLower(logs), "deprecated") {
```

Comment:
I like the idea of checking for deprecated configuration options. However this implementation has a potential to be a very flaky check. For example prometheus was emitting deprecation message for ingress resource for few versions and there was no newer prometheus version we could update to. I feel such problems might surface also in the future.

Reply:
Yeah, that's exactly what we're seeing with the failed test on kube-state-metrics 😕 

I don't see any other way to catch this though... any ideas? If not I guess we'll have to pay extra attention when merging those automated version upgrades